cf-app/src/main/kotlin/space/crowdforce/controllers/ProjectController.kt

```diff
@@ -169,6 +179,75 @@ class ProjectController(
         activityService.deleteParticipant(userId, activityId)
     }
 
+    @GetMapping("/{projectId}/goals")
+    suspend fun getGoals(@PathVariable("projectId") projectId: Int): List<GoalUI> = goalService.findGoals(projectId)
+        .map { mapperService.map(it) }
+
+    @GetMapping("/{projectId}/goals/{goalId}")
+    suspend fun getGoal(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("goalId") goalId: Int
+    ): ResponseEntity<GoalUI> = goalService.findGoal(goalId)?.let { ResponseEntity.ok().body(mapperService.map(it)) }
+        ?: ResponseEntity.notFound().build()
+
+    @PutMapping("/{projectId}/goals/{goalId}")
+    suspend fun updateGoals(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("goalId") goalId: Int,
+        @RequestBody goalFormUI: GoalFormUI,
+        principal: Principal?
+    ): ResponseEntity<Any> {
+        return try {
+            val userId = principal?.let { userService.getUserIdByName(it.name) } ?: throw UnauthorizedAccessException()
+
+            goalService.updateGoal(projectId, goalId, userId, goalFormUI.name, goalFormUI.description, goalFormUI.progress)
+
+            ResponseEntity.ok().build()
+        } catch (e: UnauthorizedAccessException) {
+            ResponseEntity.status(HttpStatus.FORBIDDEN).body(e.message)
+        } catch (e: OperationException) {
+            ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.message)
+        }
+    }
+
+    @DeleteMapping("/{projectId}/goals/{goalId}")
+    suspend fun deleteGoals(
+        @PathVariable("projectId") projectId: Int,
+        @PathVariable("goalId") goalId: Int,
+        principal: Principal?
+    ): ResponseEntity<Any> {
+        return try {
+            val userId = principal?.let { userService.getUserIdByName(it.name) } ?: throw UnauthorizedAccessException()
+
+            goalService.deleteGoal(projectId, goalId, userId)
+
+            ResponseEntity.ok().build()
+        } catch (e: UnauthorizedAccessException) {
+            ResponseEntity.status(HttpStatus.FORBIDDEN).body(e.message)
+        } catch (e: OperationException) {
+            ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.message)
+        }
+    }
+
+    @PostMapping("/{projectId}/goals")
+    suspend fun addGoals(
+        @PathVariable("projectId") projectId: Int,
+        @RequestBody goalFormUI: GoalFormUI,
+        principal: Principal?
+    ): ResponseEntity<Any> {
+        return try {
+            val userId = principal?.let { userService.getUserIdByName(it.name) } ?: throw UnauthorizedAccessException()
+
+            val goal = mapperService.map(goalService.addGoal(projectId, userId, goalFormUI.name, goalFormUI.description, goalFormUI.progress))
+
+            ResponseEntity.ok().body(goal)
+        } catch (e: UnauthorizedAccessException) {
+            ResponseEntity.status(HttpStatus.FORBIDDEN).body(e.message)
```

Comment:
It would be better to map the exceptions on `ExceptionHandler` level:

```
@ExceptionHandler(EntityNotFoundException.class)
   protected ResponseEntity<Object> handleOperationException
...
```